Tag outputs (#256)
stitchfix/hamilton: decorators.md

      and resolves to
    `True` or `False`. You probably don't want to do this.

## @tag and friends

### @tag

> Allows you to attach metadata to a node (any node decorated with the function).

Reviewer: ```suggestion
Allows you to attach metadata to an output(s), i.e. all nodes generated by a function and its decorators.
```

def final_column(intermediate_column: pd.Series) -> pd.Series:
    pass
```
### @tag_outputs

`tag_outputs` enables you to attach metadata to a function that outputs multiple nodes,

> and give different tag values to different items:

Reviewer: ```suggestion
and give different tag values to different outputs:
```
items isn't a term we use.

```python
import pandas as pd
from hamilton.function_modifiers import tag_outputs, extract_columns

def intermediate_column() -> pd.Series:
    pass

@tag_outputs(
    public={'accessibility' : 'public'},
    private={'accessibility' : 'private'})

> @extract_columns('public', 'private')
> def data_used_in_multiple_ways() -> pd.DataFrame:
>     return load_some_data(...)
> ```
> 
> In this case, the tag `accessibility` would have different values for the two nodes produced by the `data_used_in_multiple_ways`
> function -- `private` for `private` and `public` for `public`.

Reviewer: ```suggestion
@extract_columns('column_a', 'column_b')
def data_used_in_multiple_ways() -> pd.DataFrame:
    return load_some_data(...)
    ```

In this case, the tag `accessibility` would have different values for the two nodes produced by the `data_used_in_multiple_ways`
function -- `public` for `column_a` and `private` for `column_b`.
```
It's confusing to have columns and tags be named the same. So changed to make it clearer what is what.

Reviewer: note I don't know how to escape the backticks here in github...

> Note that you can use `@tag_outputs` to override `@tag` if you so desire. For that to work,
> you would put `tag_outputs` above `tag` (as it would be applied last).

Reviewer: Need to provide context for this.
```suggestion
A note on decorator precedence. If using `@tag` together with`@tag_outputs` on a function (you might want to do this because you use `@tag` to "tag" all nodes with a certain set of values, and `@tag_outputs` to "tag" specific outputs with specific values), they will be applied in order up from the function. So if you desire to override `@tag`, for that to work, you would put `tag_outputs` above `tag` (as it would be applied last).
```

```python
import pandas as pd
from hamilton.function_modifiers import tag_outputs, tag, extract_columns

def intermediate_column() -> pd.Series:
    pass

@tag_outputs(
    public={'accessibility' : 'public'},
    private={'accessibility' : 'private', 'common_tag' : 'bar'})
@tag(common_tag="foo")

> @extract_columns('public', 'private')
> def data_used_in_multiple_ways() -> pd.DataFrame:
>     return load_some_data(...)
> ```
> 
> In the case above, `common_tag` would resolve to `foo` for `public` and `bar` for `private`.
> Attempting an override in the reverse direction is currently undefined behavior.

Reviewer: ```suggestion
@extract_columns('column_a', 'column_b')
def data_used_in_multiple_ways() -> pd.DataFrame:
    return load_some_data(...)
    ```

In the case above, `common_tag` would resolve to `foo` for `column_a` and `bar` for `column_b`.
Attempting an override in the reverse direction is currently undefined behavior.
```
